llvm/lib/SYCLLowerIR/GlobalOffset.cpp

```diff
@@ -37,396 +31,363 @@ ModulePass *createGlobalOffsetPass();
-    return {NewFunc, ImplicitOffset};
-  }
-
-  static DenseMap<Function *, MDNode *>
-  validateKernels(Module &M, SmallVectorImpl<KernelPayload> &KernelPayloads) {
-    SmallPtrSet<GlobalValue *, 8u> Used;
-    SmallVector<GlobalValue *, 4> Vec;
-    collectUsedGlobalVariables(M, Vec, /*CompilerUsed=*/false);
-    collectUsedGlobalVariables(M, Vec, /*CompilerUsed=*/true);
-    Used = {Vec.begin(), Vec.end()};
-
-    auto HasUseOtherThanLLVMUsed = [&Used](GlobalValue *GV) {
-      if (GV->use_empty())
-        return false;
-      return !GV->hasOneUse() || !Used.count(GV);
-    };
-
-    llvm::DenseMap<Function *, MDNode *> EntryPointMetadata;
-    for (auto &KP : KernelPayloads) {
-      if (HasUseOtherThanLLVMUsed(KP.Kernel))
-        llvm_unreachable("Kernel entry point can't have uses.");
-
-      EntryPointMetadata[KP.Kernel] = KP.MD;
+    for (Function::arg_iterator FuncArg = Func->arg_begin(),
+                                FuncEnd = Func->arg_end(),
+                                NewFuncArg = NewFunc->arg_begin();
+         FuncArg != FuncEnd; ++FuncArg, ++NewFuncArg) {
+      FuncArg->replaceAllUsesWith(NewFuncArg);
     }
 
-    return EntryPointMetadata;
-  }
+    // Clone metadata of the old function, including debug info descriptor.
+    SmallVector<std::pair<unsigned, MDNode *>, 1> MDs;
+    Func->getAllMetadata(MDs);
+    for (auto MD : MDs)
+      NewFunc->addMetadata(MD.first, *MD.second);
 
-  virtual llvm::StringRef getPassName() const override {
-    return "Add implicit SYCL global offset";
+    ImplicitOffset = std::prev(NewFunc->arg_end());
+  }
+  assert(ImplicitOffset && "Value of implicit offset must be set.");
+
+  // Add bitcast to match the return type of the intrinsic if needed.
+  if (ImplicitArgumentType != ImplicitOffsetPtrType) {
+    BasicBlock *EntryBlock = &NewFunc->getEntryBlock();
+    // Make sure bitcast is inserted after alloca, if present.
+    BasicBlock::iterator InsertionPt =
+        ImplicitOffsetAllocaInserted
+            ? std::next(cast<AllocaInst>(ImplicitOffset)->getIterator())
+            : EntryBlock->getFirstInsertionPt();
+    IRBuilder<> Builder(EntryBlock, InsertionPt);
+    ImplicitOffset = Builder.CreateBitCast(
+        ImplicitOffset,
+        Type::getInt32Ty(M.getContext())->getPointerTo(TargetAS));
   }
 
-private:
-  // Keep track of which functions have been processed to avoid processing twice
-  llvm::DenseMap<Function *, Value *> ProcessedFunctions;
-  // Keep a map of all entry point functions with metadata
-  llvm::DenseMap<Function *, MDNode *> EntryPointMetadata;
-  llvm::Type *KernelImplicitArgumentType;
-  llvm::Type *ImplicitOffsetPtrType;
-
-  TargetHelpers::ArchType AT;
-  unsigned TargetAS = 0;
-};
+  ProcessedFunctions[NewFunc] = ImplicitOffset;
+
+  // Return the new function and the offset argument.
+  return {NewFunc, ImplicitOffset};
+}
 
-} // end anonymous namespace
+DenseMap<Function *, MDNode *> GlobalOffsetPass::validateKernels(
```

Comment:
This function does two tasks: 1. validate kernels and 2. extract all entry point metadata. For readability sake, it could be useful be split it into two separate functions, one for each of the above tasks. Thanks

Reply:
I agree that the naming might be a bit misleading, how about I rename it to `generateKernelMDNodeMap` and update the comments. Splitting the function to do essentially one operation over the same loop of kernels seems a bit wasteful.